Finish first batch of getting started tutorial changes (#10579)
wagtail/wagtail: docs/getting_started/tutorial.md

# Your first Wagtail site

This tutorial shows you how to build a blog using Wagtail. This tutorial gives you hands-on experience with some of Wagtail's features.

> It's recommended that you understand the fundamentals of Python programming and the use of the Django framework in order to easily follow this tutorial.

Reviewer: How about

```suggestion
A basic knowledge of Python programming and the Django framework will help you follow this tutorial.
```

Author: I will seek a way to refactor this and avoid the use of the phrase "in order to." For the sake of  consistency, we agreed to adopt the Google developer documentation style guide. The style guide strongly advice against using "will" and it's past tense, "would". You can read more about this on the Google developer documentation style guide's [Word list](https://developers.google.com/style/word-list)

Wagtail supports Python 3.7, 3.8, 3.9, 3.10, and 3.11.

To check if you have an appropriate version of Python 3, run the following commmand:

```sh
python --version

# or

python3 --version

# or

py --version
```

> If any of the command does not return a version number or returns a version lower than 3.7, then [install Python 3](https://www.python.org/downloads/).

Reviewer: ```suggestion
If these commands do not return a version number, or return a version lower than 3.7, then [install Python 3](https://www.python.org/downloads/).
```

```{note}
Before installing Wagtail, it's necessary to install the **libjpeg** and **zlib** libraries, which provide support for working with JPEG, PNG, and GIF images through the Python **Pillow** library.
The way to do this varies by platform. See Pillow's
[platform-specific installation instructions](https://pillow.readthedocs.io/en/stable/installation.html#external-libraries).

> ```

Reviewer 2: @thibaudcolas I thought we switched to Willow though for managing images anyway? https://docs.wagtail.org/en/stable/releases/4.2.html#maintenance

I say that as long as any dependency can be automatically installed, it's not worth calling out specific dependencies for installation. If any libraries need to be installed through `homebrew` or another platform system, then we should flag them at the beginning of the tutorial.

We recommend using a virtual environment, which isolates installed dependencies from other projects.
This tutorial uses [`venv`](https://docs.python.org/3/tutorial/venv.html), which is packaged with Python 3.

**On Windows** (cmd.exe), run the following commands:

```doscon
py -m venv mysite\env

# then

> 

Reviewer 3: Ah I see it is, since we both have sequential commands and a choice here.

mysite\env\Scripts\activate.bat

# if mysite\env\Scripts\activate.bat doesn't work, run:

mysite\env\Scripts\activate
```

**On GNU/Linux or MacOS** (bash):

```sh
python -m venv mysite/env

# then
source mysite/env/bin/activate
```

**For other shells** see the [`venv` documentation](https://docs.python.org/3/library/venv.html).

```{note}
If you're using version control such as git, then `mysite` is the directory for your project.
You must exclude the `env` directory from any version control.
```

For further help on how to set up your virtual enviroment, read the instructions on MDN's [Setting up a Django development environment](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/development_environment). 

> 

Author: I added the link as a reference for further reading. Do you suggest removing it?

```sh
python -m venv mysite/env

# then
source mysite/env/bin/activate
```

**For other shells** see the [`venv` documentation](https://docs.python.org/3/library/venv.html).

```{note}
If you're using version control such as git, then `mysite` is the directory for your project.
You must exclude the `env` directory from any version control.
```

> For further help on how to set up your virtual enviroment, read the instructions on MDN's [Setting up a Django development environment](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/development_environment).

Author: I will look into this to see if we can get a more appropriate reference.

If you're using version control such as git, then `mysite` is the directory for your project.
You must exclude the `env` directory from any version control.
```

> ### Install Wagtail

Reviewer 3: ```suggestion

### Install Wagtail
```

You must exclude the `env` directory from any version control.
```

> 
> For further help on how to set up your virtual enviroment, read the instructions on MDN's [Setting up a Django development environment](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/development_environment).
> 

Reviewer 3: Discussed with @activus-d, removing this for now:

```suggestion

```

<!-- Generated with: tree -a -L 1 -F -I env mysite -->

```{note}

> Generally, in Wagtail, a single app represents a page type or content type. However, different apps can be aware of each other and access each other's data. You must register all apps within the `INSTALLED_APPS` section of the `base.py` file in the `mysite/settings` directory. Look at this file to see how the `start` command lists them in there.

Reviewer 3: Discussed with @activus-d, we’ll revert this and think more about what this note is useful for / what to do with it.

```suggestion
Generally, in Wagtail, each page type, or content type, is represented by a single app. However, different apps can be aware of each other and access each other's data. All of the apps need to be registered within the `INSTALLED_APPS` section of the `base.py` file in the `mysite/settings` directory. Look at this file to see how the `start` command has listed them in there.
```

pip install -r requirements.txt
```

This ensures that you have the relevant versions of Wagtail, Django, and any other dependencies for the project that you create.
The `requirements.txt` file contains all the dependencies needed in order to run the project.

### Create the database

> If you are yet to update the project settings, then your database is a SQLite database file in the project directory.

Reviewer: I prefer the original

Reviewer 3: Discussed with @activus-d, new version:

```suggestion
By default, your database is SQLite. To match your database tables with your project's models, run the following command:
```

```sh
python manage.py migrate
```

This command ensures that the tables in your database matches the models in your project. Every time you alter your model, for example, if you add a field to a model, then you must run the `python manage.py migrate` command to update your database.

```{note}
If you wish to deploy your Wagtail site into production, then SQLite is not recommeded. This is because of its limited scalability and concurrency control, which makes it unsuitable for large-scale applications.
```

### Create an admin user

```sh
python manage.py createsuperuser
```

This will prompt you to create a new admin user account with full permissions. It's important to note that for security reasons, the password text won’t be visible while typing.

### Start the server

```sh
python manage.py runserver
```

> If everything works, <http://127.0.0.1:8000> shows you a welcome page:

Author: The Google developer documentation greatly advocates for writing documentation in present tense. This is just an implementation of the style guide. It helps with consistency in the tone of the documentation

![Browser screenshot of "Welcome to your new Wagtail site!" page, with teal egg above the title, and links to different resources. The page is shown inside a browswer tab, with browser URL bar at the top](../_static/images/tutorial/tutorial_1.png)

```{note}
This tutorial uses `http://127.0.0.1:8000` as the URL for your development server but depending on your setup, this could be a different IP address or port. Please read the console output of `manage.py runserver` to determine the correct url for your local site.
```

> You can now access the [Admin Interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface) by logging into <http://127.0.0.1:8000/admin> with the username and password that you entered while creating an admin user through your command line terminal.

Author: Yes. It is very important. It helps with consistency and ensuring that the capitalization of our official names/concepts are in sync. For more information, read the [capitalization](https://developers.google.com/style/capitalization) section of the Google developer documentation style guide.

> You can now access the [Admin Interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface) by logging into <http://127.0.0.1:8000/admin> with the username and password that you entered while creating an admin user through your command line terminal.

Author: It is also important that we provide links to relevant official names/concepts. This will help our new users quickly adjust to and understand the various names employed within the User guide documentation. It also helps maintain consistency with the names.

> You can now access the [Admin Interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface) by logging into <http://127.0.0.1:8000/admin> with the username and password that you entered while creating an admin user through your command line terminal.

Reviewer 3: We’ve discussed today and would prefer for this to avoid Title Case as this is a generic term (e.g. Django admin interface).

![Screenshot of Wagtail’s dashboard, with "Welcome to the mysite Wagtail CMS" heading, 1 page, 0 images, 0 documents. Underneath is a "Your most recent edits" section, with the Home page listed](../_static/images/tutorial/tutorial_2.png)

## Extend the HomePage model

> Out of the box, the **home** app defines a blank `HomePage` model in `models.py`, along with a migration that creates a homepage and configures Wagtail to use it.

Reviewer 4: If we are going to bold the app name here we should do it above also.

## Extend the HomePage model

> Out of the box, the **home** app defines a blank `HomePage` model in `models.py`, along with a migration that creates a homepage and configures Wagtail to use it.

Reviewer 3: Dropping this after review with @activus-d:

```suggestion
Out of the box, the "home" app defines a blank `HomePage` model in `models.py`, along with a migration that creates a homepage and configures Wagtail to use it.
```

  Applying home.0003_homepage_body... OK
```

You can now edit the homepage within the Wagtail [Admin interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface) (on the side bar go to **Pages** and click edit beside **Homepage**) to see the new body field.

![Screenshot of an extended side panel of the Wagtail admin interface with the edit button of the home page marked in red](../_static/images/tutorial/tutorial_11.png)

Enter the text "Welcome to our new site!" into the body field, and publish the page by selecting **Publish** at the bottom of the page editor, rather than **Save Draft**.

You must update the page template to reflect the changes made
to the model. Wagtail uses normal Django templates to render each page
type. By default, it looks for a template filename formed from the app and model name,
separating capital letters with underscores. For example, HomePage within the 'home' app becomes
`home/home_page.html`. This template file can exist in any location that 

> [Django's template rules](https://docs.djangoproject.com/en/stable/intro/tutorial03/#write-views-that-actually-do-something) recognize. Conventionally, you can place it within a `templates` folder within the app.

Reviewer 4: ```suggestion
[Django's template rules](https://docs.djangoproject.com/en/stable/intro/tutorial03/#write-views-that-actually-do-something) recognise. Conventionally, you can place it within a `templates` folder within the app.
```

    "wagtail.sites",
    "wagtail.users",
    # […]

> ]

Reviewer 4: Maybe adjust the previous line also, instead of 

```
    # [...]
```

It could be

```
    #... other packages
```

```

Most of this should be familiar from our previous work with the `home_page.html` template, but we'll explain `get_children` a bit later.

> If you have a Django background, then you can notice that the `pageurl` tag is similar to Django's `url` tag, but takes a Wagtail Page object as an additional argument.

Reviewer 4: ```suggestion
If you have a Django background, then you may notice that the `pageurl` tag is similar to Django's `url` tag, but takes a Wagtail Page object as an additional argument.
```

Or

```suggestion
If you have a Django background, then you will notice that the `pageurl` tag is similar to Django's `url` tag, but takes a Wagtail Page object as an additional argument.
```

> Now you can create a new page. Here is how you can create one from the Wagtail admin interface: 

Reviewer 4: ```suggestion
Now that this is complete, here is how you can create one from the Wagtail admin interface: 
```

Not sure on these exact words but the current line feels a bit clunky.

1. Go to `http://127.0.0.1:8000/admin` and sign in with your admin user details.
2. In the Wagtail [Admin interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface), go to Pages, then click Home.
3. Add a child page to the Home page by clicking **...** at the top of the screen and selecting the option **Add child page**.
4.  Choose **Blog index page** from the list of the page types.

> 5.  Use "Our Blog" as your page title, make sure it has the slug "blog" on the Promote tab, and publish it.

Reviewer 4: Prettier will autoformat these numbers, best to run on the MD files so that indents are aligned.

Reviewer 3: Perhaps we should stop trying to follow Prettier formatting for docs, since we disabled it there on purpose? As long as the Markdown renders correctly, it seems to me we’d be better off not worrying about formatting if it can’t be done automatically.

If we do want things to be correctly formatted, then we might as well set up Prettier for Markdown.

Reviewer 4: I do think it's important, it's a bit annoying with list indents but it comes with a lot of benefits for common typos I've seen such as line breaks after/before headings, malformed backticks for code snippets, double spaces or extra spaces around words and sentences. 

Additionally, it picks up on the rst double backticks for inline code formatting and malformed sub-nesting of lists that seem ok in the editor but didn't work when rendered.

Happy to give up on nice formatting as a goal though, I know it does add friction. Every time I've done a pass on the formatter (every few months) I've found some kind of bug though. 

Previous discussions with comments/links - https://github.com/wagtail/wagtail/discussions/8753

Reviewer 4: I've raised a PR to enable Prettier - can get some feedback there. https://github.com/wagtail/wagtail/pull/10597

Reviewer 3: > It picks up on the rst double backticks 

❤️

You can now access the url, `http://127.0.0.1:8000/blog` on your site. This gives you an error page showing "TemplateDoesNotExist" because you are yet to create a template for the new page. Also, note how the slug from the Promote tab defines the page URL.

Now create a model and template for your blog posts. Edit `blog/models.py` to include:

```python
from django.db import models

from wagtail.models import Page
from wagtail.fields import RichTextField
from wagtail.admin.panels import FieldPanel

> # add this:

Reviewer 4: Note. There are a few times you have put 'add this' , maybe it's worth using the advanced line highlighting feature for some of these code snippets. 

It's a bit fiddly but really nice when wanting to call out 'diffs' for progressively adding to an existing file. 

See an example here. 

https://github.com/wagtail/wagtail/blob/401afe7a8170747ad687801b4c661eeac40a98a7/docs/reference/streamfield/blocks.md?plain=1#L546

Note. It has to be inside an rst block.

Reviewer 4: Here's the docs for the Myst parser version of this.
https://myst-parser.readthedocs.io/en/v0.16.1/syntax/syntax.html?highlight=emphasize-lines#parameterizing-directives

Double check but this may mean you do not have to reach for an RST block straight away.

````
```{code-block} python
:lineno-start: 10
:emphasize-lines: 1, 3

a = 2
print('my 1st line')
print(f'my {a}nd line')
```
````

Author: ```{code-block} python
:lineno-start: 10
:emphasize-lines: 1, 3

a = 2
print('my 1st line')
print(f'my {a}nd line')
```
I already implemented the rst block but I tried this and it works too.

Reviewer 3: I’ve decided to proceed with the "add this"-style comments for now. I like the idea of the highlighting but:

- If it’s RST-only, this seems like a dealbreaker to me
- Even if it works in Markdown, it’s only based on background color, which wouldn’t be enough for us to meet our accessibility goals.

Author: @thibaudcolas @lb-  Should I implement this with the Myst parser then? I will remove the RST blocks and use the "add this"-style comments for now. I can always implement either the Myst parser or RST in a new PR once we make a final decision regarding this.

Reviewer 4: Hey. I agree with Thibaud that simpler is probably better here, an add this type comment gets the job done and it's a good call on the accessibility constraint.

We may want to add something to our writing documentation page to recommend this approach.

Note the use of Wagtail's built-in `get_parent()` method to obtain the
URL of the blog this post is a part of.

Now, go to your [Admin interface](https://guide.wagtail.org/en-latest/concepts/wagtail-interfaces/#admin-interface) and create a few blog posts as children of `BlogIndexPage` by following these steps:

> 1. Click **Pages** from the Wagtail [Side bar](https://guide.wagtail.org/en-latest/how-to-guides/find-your-way-around/#the-sidebar), and then click **Home**

Reviewer 4: For this next set of sub-steps. I recommend something like a bold small title for each. 

Similar to what you have for number three but bolded or with a heading. 

Having a mix of numbered items and content is a bit tricky with markdown and also a bit hard to read.

class BlogIndexPage(Page):
    intro = RichTextField(blank=True)

    # add the get_context method:

    def get_context(self, request):
        # Update context to include only published posts, ordered by reverse-chron
        context = super().get_context(request)
        blogpages = self.get_children().live().order_by('-first_published_at')
        context['blogpages'] = blogpages
        return context

    # ...
```

Here is a quick breakdown of the changes we made:

1. Retrieve the original context.
2. Create a custom QuerySet modifier.
3. Add it to the retrieved context.

> 4. Return the modified context back to the view.

Reviewer 4: These indents are added intentionally by the default Prettier markdown formatting, I know it's odd but its meant to be that way for larger numbers to have a consistent visual indent in the code.

There are a few new concepts here, so let's take them one at a time:

> 1. Inheriting from `Orderable` adds a `sort_order` field to the model, to keep track of the ordering of images in the gallery.

Reviewer 4: Same here with the indents, it's intentionally there and we should aim to keep them to avoid unnecessary diffs and churn in the code when we do another formatting pass later.
